packages/core-p2p/src/network-monitor.ts

```diff
@@ -301,57 +300,106 @@ export class NetworkMonitor implements P2P.INetworkMonitor {
         return { forked: true, blocksToRollback: Math.min(lastBlock.data.height - highestCommonHeight, 5000) };
     }
 
-    public async syncWithNetwork(
+    public async downloadBlocksFromHeight(
         fromBlockHeight: number,
-        maxParallelDownloads: number = 25,
+        maxParallelDownloads: number = 10,
     ): Promise<Interfaces.IBlockData[]> {
-        try {
-            const peersAll: P2P.IPeer[] = this.storage.getPeers();
-            const peersFiltered: P2P.IPeer[] = peersAll.filter(peer => !peer.isForked());
+        const peersAll: P2P.IPeer[] = this.storage.getPeers();
 
-            if (peersFiltered.length === 0) {
-                this.logger.error(
-                    `Could not download blocks: Failed to pick a random peer from our list of ${peersAll.length} peers: all are either banned or on a different chain than us`,
-                );
+        if (peersAll.length === 0) {
+            this.logger.error(`Could not download blocks: we have 0 peers`);
+            return [];
+        }
 
-                return [];
-            }
+        const peersNotForked: P2P.IPeer[] = peersAll.filter(peer => !peer.isForked());
+
+        if (peersNotForked.length === 0) {
+            this.logger.error(
+                `Could not download blocks: We have ${peersAll.length} peer(s) but all ` +
+                `of them are on a different chain than us`
+            );
+            return [];
+        }
 
-            const networkHeight: number = this.getNetworkHeight();
+        const networkHeight: number = this.getNetworkHeight();
+        const chunkSize: number = 400;
+        let chunksMissingToSync: number;
+        if (!networkHeight || networkHeight <= fromBlockHeight) {
+            chunksMissingToSync = 1;
+        } else {
+            chunksMissingToSync = Math.ceil((networkHeight - fromBlockHeight) / chunkSize);
+        }
+        const chunksToDownload: number = Math.min(chunksMissingToSync, peersNotForked.length, maxParallelDownloads);
+
+        // We must return an uninterrupted sequence of blocks, starting from `fromBlockHeight`,
+        // with sequential heights, without gaps.
+
+        let downloadJobs = [];
+        let downloadResults = [];
+        let someJobFailed: boolean = false;
+        let chunksHumanReadable: string = "";
+
+        for (let i = 0; i < chunksToDownload; i++) {
+            const height: number = fromBlockHeight + chunkSize * i;
+            const isLastChunk: boolean = i === chunksToDownload - 1;
+            const blocksRange: string = `[${height + 1}, ${isLastChunk ? ".." : height + 1 + chunkSize}]`;
+
+            downloadJobs.push(async () => {
+                let blocks: Interfaces.IBlockData[];
+                let peer: P2P.IPeer;
+
+                for (peer of shuffle(peersNotForked)) {
```

Comment:
For every chunk you start from a "shuffled" list of peers, so the 1st peer you download from can be the same for say 5 chunks : you will hit rate limit and be disconnected from the peer, causing the 5 downloads to fail (then yes I understand that you pick another peer but then you have more chances to hit rate limit again and cause some cascading download errors due to rate limit).
I'd say at least for the 1st try for every chunk to be downloaded from a different peer, then somehow be careful with rate limit for the "fallback" peers you download from if 1st one failed.

Reply:
Right, good point! Should be addressed in 21e3d0036.

"Resolve conversation" this if you are ok with it.

Reply:
yes looking good now :ok_hand: